Add optional plural name for unit and food

Knalltuete5000 opened the pull request:
Add an optional plural name for unit and food.

This is only a simple approche on adding a plural name for unit and food.

The plural name for unit and food is adjusted for the ingredientComponent based on the amount needed but only if it is stored in the database for the unit/food.

Additionaly to options are available:

-  Always show the plural unit name
-  Always show the plural food name

These options may be usefull in some languages.

vabene1111 commented:
thank you for the PR.

There has been some discussion about this here https://github.com/TandoorRecipes/recipes/issues/345 and in some other issue i cant remember. 

I will need to review the code and see which impact it has but generally think we should be able to get this included in one of the next releases.

Knalltuete5000 commented:
Thanks for the quick response. 

I know that this is a very simple solution on how to deal with such an issue.

What I would like to mention is that there is an additional migration step. This step was automatically created when the other migration steps where created. I am not sure if it is needed. If not I will remove it.

vabene1111 commented:
I probably need to recrrate the migration anyway when merging this so dont sorry about that.

Knalltuete5000 commented:
Is there anything I can do that this change/feature can be merged. Or is something blocking this, e.g. planed for a later release?

vabene1111 commented:
i wasn't sure if you were still working on this or not and also i am trying to get the current release finished before merging this. 

To be totally honest I am hesitant to introduce the complexity this feature brings as i am currently working on cleaning things up and improving the architecture and foundation but i also dont want to let your work go to waste so i will merge it with the next release. If you want to do me a favor you can rebase it after the next release and write a short comment here so i can get it merged soon after (maybe 1-2 weeks as some hotfixes will likely be needed).

sorry for the wait but it makes things much easier for me and i sadly do not have much time these days ...

Knalltuete5000 commented:
This mr is ready to merge, I was just rebasing it to the latest commits and ensure it still worked. 

Sounds good to me with next release 👍

Thanks for the effort you put into this project

Knalltuete5000 commented:
I have just rebased the branch to the latest release.

There is one question that came into my mind: Do the importers still work with this change or do they have any problems with this additional field? I have never really used them and maybe some of you have more information on this topic.

If this change  has no effect on the importers, then this pr is ready to merge.

vabene1111 commented:
first of all sorry for taking so long to get this merged, really hate to keep you working
second: thank you very much for keeping this up to date, that super awesome
third: as long as the field is not required it should not cause any issues, so if all api calls and models still work without the optional field all is fine.

vabene1111 commented:
i just wanted to get to merging this but i got to the same points as the first time that i reviewed this. Can you comment on the questions i posted above in the code ? 

can you also just delete the migration and api.ts files, i will regenerate them after merging, that would be awesome, thanks

Knalltuete5000 commented:
I have just deleted the api.ts as well as the migration for the commit/MR.

> Can you comment on the questions i posted above in the code ?

I have not found the comments in the code (Maybe just my fault). Can you give me a hint to the comments/questions so I can answer them and find a solution.

vabene1111 commented:
if you scroll up in this post here, the third entry is "vabene111 started a review", below that should be a section of code comments or is that only visible to me/repository members ? that would make no sense 🤔 if you dont see them let me know and i will post them some other form

Knalltuete5000 commented:
Ok. That is weird. I do not see the review you have made.

vabene1111 reviewed (commented):
i have submitted the review, do you see it now ?

StephenBrown2 commented on vue/src/components/IngredientComponent.vue:
```diff
@@ -15,12 +15,31 @@
+                    </template>
+                    <template v-else>
+                        <span v-if="ingredient.always_use_plural_unit">{{ ingredient.unit.plural_name}}</span>
+                        <span v-else-if="(ingredient.amount * this.ingredient_factor) > 1">{{ ingredient.unit.plural_name }}</span>
```
Seconded for english, values less than one are also often pluralized, so the comparison should be `!= 1` for plural usage.

tomtjes replied:
it's 1/2 cup honey though...

vabene1111 replied:
i guess this comes down to two options

1. not implement pluralization as it simply does not work in any consistent matter across different languages (as we did until now, i personally just use things like `Möhre(n)` in my database
2. implement it the way it is right now in this PR and live with the limitations

I am really not sure and would love some input on this, i am kinda leaning towards leaving out pluralization because
1. the added complexity from a technical standpoint
2. the added complexity to the user (and userinterface)
3. the limitation that we cannot implement it correctly without adding even more complexity
4. many users likely not even using the feature as it is a lot of work to enter all that data

that said for some people who want to have well structured databases its the only way to actually enter data correctly. 

really looking for some feedback here

Also here is an interesting article about how wrong you can be with assumptions about languages and grammar even for something as simple as a name https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/

Knalltuete5000 replied:
Sorry for the late response.

I have thought about the topic for a while now and there are maybe only these two options you have mentioned. It is a difficult topic and my information is just from knowing 1 and a half languages.

What do you think about adding a setting, either server-wide (toggled by an administrator) or per space/user to use this kind of pluralization (by default disabled).

This will not remove any complexity from the UI maybe it makes it even worse.

If this is not the solution you are looking for feel free to either provide an other solution or to reject the MR. I am fine with both results.

vabene1111 replied:
i am sorry for not working on this earlier as well, i will be on vacation for the next few weeks and once i have time afterwards i will look into this again, sorry.

Knalltuete5000 commented on cookbook/models.py:
```diff
@@ -590,6 +592,8 @@ class Ingredient(ExportModelOperationsMixin('ingredient'), models.Model, Permiss
     is_header = models.BooleanField(default=False)
     no_amount = models.BooleanField(default=False)
+    always_use_plural_unit = models.BooleanField(default=False)
+    always_use_plural_food = models.BooleanField(default=False)
```
I have thought about putting it in the food/unit model but I think in generell it is better to decide it on an "ingredient-basis" if you want to always use the plural form.

To always use the plural for of a unit I do not have an example but there might  be languages where you need it. Correct me if I am wrong.
The use of a plural for the food can vary, e.g. in the German language

1 Bund Möhren  | 1 g Möhre
2 Bund Möhren  | 100 g Möhren

Even if this example is not the best there might be situations where you cannot say you want to use always the plural form of the food.

vabene1111 replied:
ok that example makes sense, i hate the extra complexity but i guess then its needed

Knalltuete5000 commented on cookbook/serializer.py:
```diff
@@ -438,6 +443,9 @@ def get_substitute_onhand(self, obj):
 
     def create(self, validated_data):
         name = validated_data.pop('name').strip()
+        plural_name = validated_data.pop('plural_name', None)
```
I am not sure what you mean with the "transparent pluralization". Can you give a short example on how it should behave.

If I understand the labels correct they can definitely help to standardize the frontend logic. But may this be a refactoring task for the project to include  it in multiple models

vabene1111 replied:
ok everything fine, just wanted to debate this as i wasn't sure myself either

Knalltuete5000 commented:
Hey there,

I have just rebased to the lastest commit and added an option to the space to enable the usage of the plural name for units and foods. By default this feature is turned off.

This change does not reduce the complexity of the feature instead it adds more but I hope this new setting will be enough to merge this pr. Let me know what  you think about it.

I will add a documentation about the setting to the feature-docs to make clear what this feature enables, how it works and what it is not supposed to bring to the application later.

Szeraax commented:
Wow, this is a bit of a wild read. I totally hadn't considered that different languages would pluralize different amounts differently. I wonder if its common across all imperial units (UK, USA, etc.) to only use plural for amounts > 1 and if its common across metric (most everyone else) to use plural for anything != 1. 

Anyone have knowledge on the matter?

vabene1111 commented:
sorry, had to get the other fix out for a security fix, will wait a few days for the new version to settle and fixes to stop so i can update the hosted version, after that will get this merged.

Knalltuete5000 commented:
Hey there,

This will be a last attempt to get this merged... or not. The topic was discussed here quiet a bit and I'm not sure if the feature is wanted or not. If it is wanted I need to know and I am willing to put more time in this feature if there are some code reviews on how to improve the code.

More about the current situation: I have added an additional documentation for the setting. This page may come in handy for other options to be described here.

Please let me know how you think to proceed with this PR.

vabene1111 commented:
hey @Knalltuete5000 i am so sorry for leaving you hanging like this for so long. I have been super busy the last few weeks and just got back to doing the bare minimum over the weekend. 

To be 100% honest while I really love this feature to be added (and it has been requested many times) I have become a bit afraid of adding features especially from PRs as often times stuff is thrown in and I am left with the aftermath of fixing whatever arises throughout the coming releases and whats noticed by the users. 

That said I told you that I would merge it and you have put so much work into this so I will honor my promise and get this merged. I will try to get it in as soon as possible!

Knalltuete5000 commented:
Thank you your honst answer. 

Please let me know how I can support you. 
And, if it is merged and there is a problem with this feature, please let me know as well. I will then try my best to help where I can.

vabene1111 commented:
ok so sorry for the long wait, finally got around to getting this reviewed. I opened #2181 as an improvement issue for the future but other than that its all fine. 

One thing just so you know: you do not need to update all translation files. Just `en.json` is needed, everything else is done automatically by the translation system over at translate.tandoor.dev

vabene1111 commented:
thanks for all the patience, the well written code and PR and the nice feature :)

vabene1111 commented:
hey @Knalltuete5000 i am a bit confused. I cant imagine you forgot that but I just added the use_plural parameter to the RecipeView and sub components (step/ingredient). Did you have some different solution for that?

When i tested it it did not show plurlars in the view 

see 39ccf7bbcf5758d45d246dc53d61097f2147f456

Knalltuete5000 commented:
Thank you!

Because of the `use_plural`: either I forgot it (what I could imagine) or it was removed during one of my rebases and I did not notice it. Thanks for the quick reaction.

vabene1111 commented:
ok perfect, just wanted to make sure that i did not circumvent any other logic that you placed somewhere and that is now doubled. Thanks for the quick response :)
